### DISCUSSION TITLE
Attention Mask for Packed Sequences (via Attention Bias)

### DISCUSSION THREAD

**Discussion by xingyaoww** [erfanzar/EasyDeL#139] [Announcements]
Hi @erfanzar,

Thanks for the great repo! It looks really useful for training open-source models on TPU and GPUs!

I wonder if it is easy to implement the feature that allows users to pass in packed sequences (see [this](https://github.com/Dao-AILab/flash-attention/issues/432#issuecomment-1698610752)) that allows us to maximize the hardware utilization?

The idea is that the user can provide a list of `attention_mask_in_length`, then under the hood, we tweak the `attention_mask` OR adjust the attention bias accordingly, so that two different examples cannot attend to each other.

![](https://private-user-images.githubusercontent.com/16218783/264256249-2f51a6ea-10df-4f4d-bbf2-928dd4e919f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzNDU1NTAsIm5iZiI6MTcxMDM0NTI1MCwicGF0aCI6Ii8xNjIxODc4My8yNjQyNTYyNDktMmY1MWE2ZWEtMTBkZi00ZjRkLWJiZjItOTI4ZGQ0ZTkxOWY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDE1NTQxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThjNjI2ZTRjNmQzYTRkZjJkNTc0N2U4MmU5ODY1MzcxOGRiNmFlYjhiYjg5NWVkNTYyYTY1Yjk5YzczNzNkZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2q6qHMLP6CiatJrsRFDamHP2YfPiivBl_hg4r4bKNvw)

I see that you support attention bias for flash attention [here](https://github.com/erfanzar/EasyDeL/blob/main/lib/python/EasyDel/modules/easy_attention.py#L193), which could be a good starting point (e.g., just set the attention across sequences to be -inf). However, I find that the [`flash_func`](https://github.com/erfanzar/EasyDeL/blob/main/lib/python/EasyDel/modules/easy_attention.py#L422-L430) has a different function signature:
1. [For GPU](https://github.com/erfanzar/EasyDeL/blob/f5be48bd0e5f5b211c7ce06ccb91d6687c92af8a/lib/python/EasyDel/modules/easy_attention.py#L37), `flash_func` is `flash_attn_gpu.mha`, which does not really has the argument `bias` according to [here](https://github.com/erfanzar/FJFormer/blob/da9a470e287cda56116e63a0cf3380692cb8c9b9/src/fjformer/pallas_operations/flash_attention/gpu/jax_flash_attn_gpu.py#L172).
2. While [For TPU] (`flash_attn_tpu.flash_attention`), [that argument exists](https://github.com/erfanzar/FJFormer/blob/da9a470e287cda56116e63a0cf3380692cb8c9b9/src/fjformer/pallas_operations/flash_attention/tpu/jax_flash_attn_tpu.py#L143).

Is this intended? If not, do we need to fix the GPU's implementation to be the same as TPU's? Is this the best way to implement such packing feature?

**Comment by erfanzar** [comment]
Hello, and thanks for using EasyDeL.

It would be a cool feature to add is there any code related to someone who has implemented this before or not?

for GPU flash attention I want to use JAX-triton to port flash attention2 implementation to jax-Pallas and use that it would be an easier, faster, and better option to use that instead of re-creating flash attention 2 for Pallas jax GPUs.

is there any related paper released for additional information on implementation?, as I understood from what has been explained here it's not that much of a hard thing to implement we just need a 3D attention mask or causal mak which is not hard to implement for GPUs, but for TPUs there has to be some little tricks for sharding and multi-hosting implementation.

**Comment by xingyaoww** [comment]
Hi @erfanzar ,

Thanks for your response! I think using Jax-triton is a great idea!!

Regarding specific implementation for packed attention: [Here](https://github.com/xingyaoww/Megatron-LLM/blob/1b07577760437c61189f949643a52f49ba2f2a7e/megatron/model/transformer.py#L540-L573) is what I've implemented for Megatron-LLM based on the [issue i shared earlier](https://github.com/Dao-AILab/flash-attention/issues/432#issuecomment-1698610752).

However, I don't know if it will be super helpful since it directly uses flash-attention's `flash_attn_varlen_qkvpacked_func`. I quickly search over the flash attention repo and find the implementation of this function: [here](https://github.com/Dao-AILab/flash-attention/blob/26c9e82743e5f65cab4e09898995cac70c3de267/flash_attn/flash_attn_interface.py#L839-L887) and [here](https://github.com/Dao-AILab/flash-attention/blob/26c9e82743e5f65cab4e09898995cac70c3de267/flash_attn/flash_attn_interface.py#L271-L311). That function seems to rely on `varlen_fwd` in [their CUDA kernel implementation](https://github.com/Dao-AILab/flash-attention/blob/26c9e82743e5f65cab4e09898995cac70c3de267/csrc/flash_attn/flash_api.cpp#L496), which would be too low level i feel.

We only have [`fused_attention_kernel` function in jax-triton](https://github.com/jax-ml/jax-triton/blob/1999d9b116bf7c5c94f70de4a45b414255366fbe/examples/fused_attention.py#L35), not sure if we could repurpose that function in a way to implement this cleverly.

Also, I wonder whether you have an estimated timeline for integrating jax-triton's attention into EasyDel? I might also be able to take a look at this if you integrated jax-triton into the repo :-)!

**Comment by erfanzar** [comment]
I'll work on that ASAP and actually, I am interested in working on this since there are available CUDA and triton implementations
I'll first start with implementing that on GPU via jax-triton and then create a Pallas version for that.
and I'll be happy to integrate JAX-triton attention and other low-level GPU functions into EasyDeL for sure, but it would take some time right now I have to debug the created flash and ring attention created for TPUs and implement the SFTTrainer and after that, I can work on this but right now my priority is to fix the upcoming issue, but I guess I'll start integrating JAX-triton to FJFormer in next week and then connect FJFormer to EasyDeL
